README.rst
.. ==================================================
.. FOR YOUR INFORMATION
.. --------------------------------------------------
.. -*- coding: utf-8 -*- without BOM.

.. ==================================================
.. DEFINE SOME TEXT ROLES
.. --------------------------------------------------

.. role::   typoscript(code)

.. role::   ts(typoscript)
    :class:  typoscript

.. role::   php(code)

.. highlight:: php

.. _start:

==========================================
Netresearch Content Delivery Network Tools
==========================================

.. only:: html

    :Classification:
        nr_cdn

    :Version:
        2.0.2

    :Language:
        en

    :Description:
        This extension will link static media in your page to your Content Delivery Network
        or just any other static server for media delivery.

        http://en.wikipedia.org/wiki/Content_delivery_network

        So you can use your full featured CDN or just use some lightweight fast
        http server like Lighttpd to deliver content to your customers faster
        and offload traffic from your CMS servers.

    :Keywords:
        netresearch, cdn, aws, cloudfront

    :Copyright:
        2010-2014

    :Author:
        Alexander Opitz, Sebastian Mendel

    :Email:
        alexander.opitz@netresearch.de, sebastian.mendel@netresearch.de

    :License:
        This document is published under the Open Content License
        available from http://www.opencontent.org/opl.shtml

    :Rendered:
        |today|


.. contents:: Table of Contents

What does it do
===============

This extension will link static media in your page to your Content Delivery Network
or just any other static server for media delivery.

http://en.wikipedia.org/wiki/Content_delivery_network

So you can use your full featured CDN or just use some lightweight fast
http server like Lighttpd to deliver content to your customers faster
and offload traffic from your CMS servers.

nr_cdn will rewrite URLs to static files in served pages e. g.:

- /fileadmin/upload/image.jpg => //cdn.example.org/fileadmin/upload/image.jpg
- /typo3conf/ext/my_ext/res/style.css => //cdn.example.org/typo3conf/ext/my_ext/res/style.css

if you want you can exclude paths from rewriting with the help of the ignore_paths option.

TypoScript Configuration
========================

Constants
---------

.. code-block:: typoscript

    CDNURL = //cdn.domain.tld
    CDN_ignoreslash = 1

Setup
-----

.. code-block:: typoscript

    # CDN URL prefix/host
    URL = {$CDNURL}

    # whether to ignore leading slashes in given replacement paths
    ignoreslash = {$CDN_ignoreslash}

    ignore_paths {
        1 {
            # paths to be excluded from replacing/prefixed with CDN host
            path = fileadmin/no_cdn
        }
        2 {
            # paths to be excluded from replacing/prefix with CDN host
            path = fileadmin/no_cdn
            ext {
                # file extension to be excluded from replacement with CDN host
                10 = .js
                # file extension to be excluded from replacement with CDN host
                20 = .png
            }
        }

    }

    paths {
        1 {
            # paths to be replaced/prefixed with CDN host
            path = fileadmin
        }
        2 {
            # paths to be replaced/prefixed with CDN host
            path = typo3temp
        }
        3 {
            # paths to be replaced/prefixed with CDN host
            path = typo3conf
            ext {
                # file extension to be replaced with CDN host
                10 = .js
                # file extension to be replaced with CDN host
                20 = .png
                # file extension to be replaced with CDN host
                30 = .gif
                # file extension to be replaced with CDN host
                40 = .jpg
            }
        }
    }

Note
----

Don't forget to clear the cache afterwards.

What won't go to your CDN:

- thumbnails, cause this will be generated through TYPO3s thump.php
- class files, so you shouldn't have problems with java/website interactions
  caused by XSS

PHP Configuration
=================

.. code-block:: php

    $GLOBALS['CDN_CONF_VARS'] = array(
        // host name for CDN
        'host' => 'cdn.example.org',

        // whether to ignore leading slashes in given replacement paths
        'ignoreslash' => true,

        // paths to ignore for replacement/prefix with CDN host
        'ignore_paths' => array(
            // ignore every file in no_cdn
            'fileadmin/no_cdn' => null,
            // ignore only static files
            'fileadmin/not_all_cdn' => array('.js', '.png', '.gif', '.jpg'),
        ),

        // paths to be replaced/prefixed with CDN host
        'paths' => array(
            'fileadmin' => null, // every file
            'typo3temp' => null, // every file
            // only static files
            'typo3conf' => array('.js', '.png', '.gif', '.jpg'),
        ),
    );



ChangeLog
2014-09-02  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 2.0.2

2014-09-02  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 2.0.1

2014-08-26  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 2.0.0

2014-07-03  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 1.2.1

2014-07-03  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 1.2.0

2014-03-14  Tobias Hein  <tobias.hein@netresearch.de>

	* TYPO-1840: remove unused feature ignoreslash -  do ignores  with the new ignore_paths option instead

2014-02-20  Tobias Hein  <tobias.hein@netresearch.de>

	* TYPO-1840: add ingore_path array to nr_cdn config -  now the  patters in ignore_path would be ignored for cdn replacement

2013-10-17  Marian Pollzien  <marian.pollzien@netresearch.de>

	* TYPO-1840: added the current http host to the regex for the cdn url replacement, so absolute urls are also replaced

2013-09-19  Tobias Hein  <tobias.hein@netresearch.de>

	* Version: 1.1.0

2013-09-13  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* TYPO-1691: add post render filter to rewrite asset URLs to CDN

2013-03-07  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 1.0.1

2013-03-01  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 1.0.0

2013-03-01  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Prefix all image path for cImage with CDN Host -  not only  the ones from fileadmin/

2012-12-11  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 0.15.0

2012-10-04  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 0.14.0

2012-08-31  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Make TypoScript configuration finally fully working

2012-08-27  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 0.13.1

	* minor coding style, doc and test improvments

2012-08-24  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 0.13.0

	* Add autoloader configuration

	* rename ux_ux_tslib_fe to ux_tslib_fe

2012-05-29  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 0.12.0

2012-05-14  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* SDM-10717: fix file allow pattern handling for CDN URL rewriting
      this fixes URLs to PHP-scripts rewritten with CDN prefix

2012-04-26  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 0.11.0

2012-03-28  Alexander Opitz  <alexander.opitz@netresearch.de>

	* SDM-9944: Update to typo46 INTincScript

	* SDM-9944: Fix API change in Typo46

2012-03-09  Alexander Opitz  <alexander.opitz@netresearch.de>

	* SDM-9831: support file extensions on path to replace
    Now you can configure to use CDN only for specific file extension

2012-02-01  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version 0.10.0

2012-01-27  Sebastian Mednel  <sebastian.mendel@netresearch.de>

	* SDM-4961: make paths served from CDN configurable

	* SDM-4961: add an option to ignore leading slashes in pathes

	* SDM-4961: refactor code

2012-01-03  Alexander Opitz  <alexander.opitz@netresearch.de>

	* Version: 0.9.7

2011-12-21  Alexander Opitz  <alexander.opitz@netresearch.de>

	* SDM-9810: Bereinigung der multiplen Einträge in arParams

2011-12-20  Alexander Opitz  <alexander.opitz@netresearch.de>

	* Version: 0.9.6

2011-12-09  Alexander Opitz  <alexander.opitz@netresearch.de>

	* SDM-8765: don't call restore if not expected

	* SDM-8765: Add to page renderer hook for header parts.

2011-09-14  Christian Weiske  <christian.weiske@netresearch.de>

	* AMTB-48: Use ux_tslib_fe from aida_common

2011-07-11 Andre Hähnel <andre.haehnel@netresearch.de>

	* SDM-4961: use parent functionality for methods

	* Version: 0.9.5

2010-04-21  Alexander Opitz  <alexander.opitz@netresearch.de>

	* SDM-5771: cssstyledcontent connector

	* Version: 0.9.4

2010-04-21  Alexander Opitz  <alexander.opitz@netresearch.de>

	* SDM-3483: and_shadowbox and wec_discussion needs the deprecated way.

	* Version: 0.9.3

2010-03-29  Alexander Opitz  <alexander.opitz@netresearch.de>

	* SDM-2160: Fix problem with auto images in typo3temp

	* Version: 0.9.2

2010-03-17  Alexander Opitz  <alexander.opitz@netresearch.de>

	* Fix XClass Name

	* Version: 0.9.1

2010-01-25  Alexander Opitz  <alexander.opitz@netresearch.de>

	* Release NR CDN for TYPO3 4.2.1


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "nr_cdn".
 *
 * Auto generated 02-09-2014 11:29
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Netresearch Content Delivery Network Tools',
	'description' => 'Images and Medias to Content Delivery Network',
	'category' => 'fe',
	'author' => 'Alexander Opitz, Sebastian Mendel',
	'author_company' => 'Netresearch GmbH & Co. KG',
	'author_email' => 'alexander.opitz@netresearch.de, sebastian.mendel@netresearch.de',
	'constraints' => array(
		'depends' => array(
			'typo3' => '4.3.0-4.6.99',
			'php' => '5.3.0-0.0.0',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'state' => 'stable',
	'version' => '2.0.2',
	'_md5_values_when_last_written' => 'a:20:{s:9:"build.xml";s:4:"687b";s:9:"ChangeLog";s:4:"7eac";s:32:"class.user_t3libpagerenderer.php";s:4:"2055";s:26:"class.ux_tslib_content.php";s:4:"c8af";s:21:"class.ux_tslib_fe.php";s:4:"84b0";s:16:"ext_autoload.php";s:4:"1612";s:12:"ext_icon.gif";s:4:"a459";s:17:"ext_localconf.php";s:4:"5811";s:14:"ext_tables.php";s:4:"9fa8";s:10:"README.rst";s:4:"e5de";s:23:"src/Netresearch/Cdn.php";s:4:"0052";s:50:"src/Netresearch/Cdn/HookCssFilelinksGetFileUrl.php";s:4:"c98d";s:20:"static/constants.txt";s:4:"1279";s:16:"static/setup.txt";s:4:"771b";s:19:"tests/bootstrap.php";s:4:"890a";s:20:"tests/dummy-cObj.php";s:4:"6d5b";s:26:"tests/PageRendererTest.php";s:4:"a0c5";s:17:"tests/phpunit.xml";s:4:"439e";s:23:"tests/UxContentTest.php";s:4:"7c16";s:29:"tests/Netresearch/CdnTest.php";s:4:"9e27";}',
);

?>